lua/DCS-gRPC/exporters/object.lua

```diff
@@ -25,31 +25,61 @@ GRPC.exporters.position = function(pos)
 end
 
 GRPC.exporters.unit = function(unit)
-  local vector = unit:getVelocity()
-
-  local heading = math.deg(math.atan2(vector.z, vector.x))
-  if heading < 0 then
-    heading = heading + 360
-  end
-
-  local speed = math.sqrt((vector.x)^2+(vector.z)^2)
-
   return {
     id = tonumber(unit:getID()),
     name = unit:getName(),
     callsign = unit:getCallsign(),
     coalition = unit:getCoalition() + 1, -- Increment for non zero-indexed gRPC enum
     type = unit:getTypeName(),
-    position = GRPC.exporters.position(unit:getPoint()),
+    transform = GRPC.exporters.transform(unit),
     playerName = Unit.getPlayerName(unit),
     groupName = Unit.getGroup(unit):getName(),
     numberInGroup = unit:getNumber(),
-    heading = heading,
-    speed = speed,
     category = unit:getGroup():getCategory() + 1, -- Increment for non zero-indexed gRPC enum
   }
 end
 
+GRPC.exporters.transform = function(object)
+    -- https://wiki.hoggitworld.com/view/DCS_func_getPosition
+  local position = object:getPosition()
+  local coords = GRPC.exporters.position(position.p)
+  local coordsNorth = coord.LLtoLO(coords.lat + 1, coords.lon)
+
+  local u = position.p.z
+  local v = position.p.x
+  local positionNorth = GRPC.exporters.vector(coordsNorth)
+  local projection_error = math.atan2(positionNorth.x - u, positionNorth.z - v)
+
+  local forward = position.x
+  local right = position.z
+  local up = position.y
+
+  local heading = math.atan2(forward.z, forward.x)
+  local headingDegrees = math.deg(math.atan2(forward.z, forward.x))
+  if headingDegrees < 0 then
+    headingDegrees = headingDegrees + 360
+  end
+
+  local velocity = object:getVelocity()
+
+  return {
+    position = coords,
+    u = u,
+    v = v,
+    orientation = {
+      pitch = math.deg(math.asin(forward.y)),
+      yaw = math.deg(heading - projection_error),
+      roll = math.deg(-math.asin(right.y)),
+      forward = GRPC.exporters.vector(forward),
+      right = GRPC.exporters.vector(right),
+      up = GRPC.exporters.vector(up),
+    },
+    velocity = GRPC.exporters.vector(velocity),
+    heading = headingDegrees,
+    speed = math.sqrt((velocity.x)^2+(velocity.z)^2),

```

Comment:
Does this really need to be moved into the transform?

I am seeing "Heading" (which way nose is pointing) and "Speed" is the basic information that will probably work for most clients. OverlordBot, for example, will only care about heading and speed.

I see the information in the `Transform` section as being the "Advanced" stuff for those small number of people who actually care about it and want to do their own calculations while other clients can ignore its existance.

Reply:
I personally need the `heading` as part of the `GetTransform` request, other than that, I'd be fine with moving heading and speed out of it. If we move them out of it, I'd probably still include them in the `GetTransform` response as follows:
```proto
Transform transform = 1;
double speed = 2;
double heading = 3;
```

Reply:
Here is another idea. How about completely flattening the `transform` into the `Unit` message, so `unit.transform.velocity` becomes `unit.velocity` and so on. I'd then still tend to keep the `GetTransform` API call to get the subset of properties of the unit that describe it in 2D and 3D space (so basically everything currently in the `transform` message). Wdyt?

Reply:
I'd like to redact my idea above. While working again on the PR, I've come to realize that I personally found it really useful that all those information are in one message/struct, as this allows them to be easily reused in different contexts (in the unit, in the dedicated method to get the transform, in the units stream).